h2o-core/src/main/java/water/fvec/Frame.java

```diff
@@ -1619,13 +1622,49 @@ public CSVStream(Chunk[] chks, String[] names, int nChunks, CSVStreamParams parm
       StringBuilder sb = new StringBuilder();
       if (names != null) {
         sb.append('"').append(names[0]).append('"');
-        for(int i = 1; i < names.length; i++)
+        for (int i = 1; i < names.length; i++)
           sb.append(_parms._separator).append('"').append(names[i]).append('"');
         sb.append('\n');
       }
       _line = StringUtils.bytesOf(sb);
       _chkRow = -1; // first process the header line
       _curChks = chks;
+      _escapedCategoricalVecDomains = escapeCategoricalVecDomains(_curChks);
+    }
+
+    /**
+     * Escapes categorical levels of vectors and puts them in a map of escaped categorical levels.
+     * Only the domains with at least one level with an escaped quote are saved. If a domain does not need
+     * any escaping, it is considered better practice to reach to the `vec.domain()` method itself and not duplicate entries
+     * in memory here.
+     *
+     * @param chunks
+     * @return A 2D array of String[][]. Elements can be null of give domain does not need escaping.
+     */
+    private String[][] escapeCategoricalVecDomains(final Chunk[] chunks) {
+      final String[][] localEscapedCategoricalVecDomains = new String[chunks.length][];
+
+      for (int i = 0; i < chunks.length; i++) {
+        final Vec vec = chunks[i].vec();
+        if (!vec.isCategorical()) continue;
+
+        final String[] originalDomain = vec.domain();
+        final String[] escapedDomain = new String[originalDomain.length];
+
+        boolean escapingRequired = false;
+        for (int level = 0; level < originalDomain.length; level++) {
+          escapedDomain[level] = escapeQuotesForCsv(originalDomain[level]);
+          escapingRequired = escapingRequired || !escapedDomain[level].equals(originalDomain[level]);
+        }
+
+        if (escapingRequired) {
+          localEscapedCategoricalVecDomains[i] = escapedDomain;
+        } else {
+          localEscapedCategoricalVecDomains[i] = null;
```

Comment:
unnecessary - but I like that it sort of explains the code below, but still unecessary.

Reply:
That's exactly why it's there, otherwise it's of course initialized to null from the very beginning. I'll get rid of my personal style and just remove it.

Reply:
Done.